### PR TITLE
Remove radard

### PR DESCRIPTION
**Pull request by haraschax**
- Add model that predicts lead trajectory, not just lead position and velocity. This allows longitudinal MPC to anticipate the future more and be more smooth
- Remove radar
- Replace handcoded cruise control logic with MPC
- Replace handcoded FCW logic with a simple threshold check
- Simplify overall architecture
- Make velocity + acceleration API with controls explicit
- Increase thresholds for cruise accel and deccel
- Remove controlsd from longitudinal tests
- Change packets to output control trajectory
- Let controls do anticipation

**Still todo**
- Reduce cpu usage of planner and controls
- Tune accel limits and cruise mpc
- Tune lead mpc costs

### CODE REVIEW

File: tools/sim/lib/can.py [hunk 1]
```diff
@@ -66,24 +63,3 @@ def can_function(pm, speed, angle, idx, cruise_button, is_engaged):
       msg.append([k, 0, b'\x00'*v, 0])
 
   pm.send('can', can_list_to_can_capnp(msg))
-
-def sendcan_function(sendcan):
```
Comment (adeebshihadeh) [inline tools/sim/lib/can.py:70]: why was this deleted?

File: selfdrive/controls/controlsd.py [hunk 2]
```diff
@@ -62,8 +61,8 @@ def __init__(self, sm=None, pm=None, can_sock=None):
       ignore = ['driverCameraState', 'managerState'] if SIMULATION else None
       self.sm = messaging.SubMaster(['deviceState', 'pandaState', 'modelV2', 'liveCalibration',
                                      'driverMonitoringState', 'longitudinalPlan', 'lateralPlan', 'liveLocationKalman',
-                                     'managerState', 'liveParameters', 'radarState'] + self.camera_packets,
-                                     ignore_alive=ignore, ignore_avg_freq=['radarState', 'longitudinalPlan'])
+                                     'managerState', 'liveParameters'] + self.camera_packets,
+                                     ignore_alive=ignore, ignore_avg_freq=['longitudinalPlan'])
```
Comment (adeebshihadeh) [inline selfdrive/controls/controlsd.py:65]:
```suggestion
                                     ignore_alive=ignore)
```
this can be removed since it's now driven by modeld

Comment (HaraldSchafer) [reply selfdrive/controls/controlsd.py:65]: ok

File: tools/sim/lib/can.py (hunk 1 above)
Comment (HaraldSchafer) [reply tools/sim/lib/can.py:70]: That was only used in longitudinal tests?

Comment (adeebshihadeh) [reply tools/sim/lib/can.py:70]: this is for carla

File: selfdrive/controls/lib/lead_mpc_lib/generator.cpp [hunk 3]
```diff
@@ -34,8 +33,8 @@ int main( )
   Function h;
   h << exp(0.3 * NORM_RW_ERROR(v_ego, v_l, d_l)) - 1;
   h << (d_l - desired) / (0.05 * v_ego + 0.5);
-  h << a_ego * (0.1 * v_ego + 1.0);
-  h << j_ego * (0.1 * v_ego + 1.0);
+  h << a_ego;
```
Comment (pd0wm) [inline selfdrive/controls/lib/lead_mpc_lib/generator.cpp:36]: Why remove this speed dependent cost? Acceleration is more uncomfortable at higher speeds.

File: selfdrive/controls/lib/longitudinal_planner.py [hunk 4]
```diff
@@ -60,158 +58,90 @@ class Planner():
...
-    longitudinalPlan.aCruise = float(self.a_cruise)
-    longitudinalPlan.vStart = float(self.v_acc_start)
-    longitudinalPlan.aStart = float(self.a_acc_start)
-    longitudinalPlan.vTarget = float(self.v_acc)
-    longitudinalPlan.aTarget = float(self.a_acc)
-    longitudinalPlan.vTargetFuture = float(self.v_acc_future)
-    longitudinalPlan.hasLead = self.mpc1.prev_lead_status
+    longitudinalPlan.radarStateMonoTime = 0
+
+    longitudinalPlan.vStart = float(self.v_desired)
```
Comment (pd0wm) [inline selfdrive/controls/lib/longitudinal_planner.py:138]: This is using values 0.05s in the future. Think all the future projection should be handled in controlsd

File: selfdrive/test/test_onroad.py [hunk 5]
```diff
@@ -20,15 +20,14 @@
 
 # Baseline CPU usage by process
 PROCS = {
-  "selfdrive.controls.controlsd": 50.0,
+  "selfdrive.controls.controlsd": 64.0,
```
Comment (pd0wm) [inline selfdrive/test/test_onroad.py:23]: Why did controlsd usage go up?

File: selfdrive/controls/lib/lateral_planner.py [hunk 6]
```diff
@@ -173,12 +173,12 @@ def update(self, sm, CP):
       # Heading cost is useful at low speed, otherwise end of plan can be off-heading
       heading_cost = interp(v_ego, [5.0, 10.0], [MPC_COST_LAT.HEADING, 0.0])
       self.libmpc.set_weights(path_cost, heading_cost, CP.steerRateCost)
-    y_pts = np.interp(v_ego * self.t_idxs[:MPC_N + 1], np.linalg.norm(d_path_xyz, axis=1), d_path_xyz[:,1])
-    heading_pts = np.interp(v_ego * self.t_idxs[:MPC_N + 1], np.linalg.norm(self.path_xyz, axis=1), self.plan_yaw)
+    y_pts = np.interp(v_ego * self.t_idxs[:LAT_MPC_N + 1], np.linalg.norm(d_path_xyz, axis=1), d_path_xyz[:,1])
```
Comment (pd0wm) [inline selfdrive/controls/lib/lateral_planner.py:176]: Try using `np_fast.interp` on arrays that aren't numpy arrays already. That can have a measurable performance improvement.

File: selfdrive/controls/lib/longitudinal_planner.py [hunk 7]
```diff
@@ -1,31 +1,29 @@
...
-from selfdrive.controls.lib.drive_helpers import V_CRUISE_MAX
+from selfdrive.controls.lib.drive_helpers import V_CRUISE_MAX, CONTROL_N
 
 LON_MPC_STEP = 0.2  # first step is 0.2s
 AWARENESS_DECEL = -0.2     # car smoothly decel at .2m/s^2 when user is distracted
 
 # lookup tables VS speed to determine min and max accels in cruise
 # make sure these accelerations are smaller than mpc limits
-_A_CRUISE_MIN_V = [-1.0, -.8, -.67, -.5, -.30]
+_A_CRUISE_MIN_V = [-1.2, -1.2, -1.2, -1.2, -1.2]
```
Comment (pd0wm) [inline selfdrive/controls/lib/longitudinal_planner.py:20]: Just remove BPs if ther are all the same

File: selfdrive/controls/lib/longitudinal_planner.py [hunk 8]
```diff
@@ -60,158 +58,90 @@ class Planner():
...
-                                       lead_1.dRel, lead_1.vLead, lead_1.aLeadK,
-                                       lead_1.yRel, lead_1.vLat,
-                                       lead_1.fcw, blinkers) and not sm['carState'].brakePressed
-    if self.fcw:
-      cloudlog.info("FCW triggered %s", self.fcw_checker.counters)
+    if not enabled or sm['carState'].gasPressed:
+      self.v_desired = v_ego
+      self.a_desired = a_ego
+
+    # Prevent divergence, smooth in current v_ego
```
Comment (pd0wm) [inline selfdrive/controls/lib/longitudinal_planner.py:93]: Is this really necessary? Is this something we should do for lateral as well then?

**Comment by pd0wm** [comment]
Will we add a configurable longitudinal actuators delay per car?

File: selfdrive/controls/lib/lead_mpc_lib/generator.cpp (hunk 3 above)
Comment (HaraldSchafer) [reply selfdrive/controls/lib/lead_mpc_lib/generator.cpp:36]: This isn't final yet, in it's current state, there is no cost on acceleration, only jerk

File: selfdrive/controls/lib/longitudinal_planner.py (hunk 4 above)
Comment (HaraldSchafer) [reply selfdrive/controls/lib/longitudinal_planner.py:138]: Yeah I agree, still need to do that.

File: selfdrive/test/test_onroad.py (hunk 5 above)
Comment (HaraldSchafer) [reply selfdrive/test/test_onroad.py:23]: Not sure, still need to look into that.

File: selfdrive/controls/lib/lateral_planner.py (hunk 6 above)
Comment (HaraldSchafer) [reply selfdrive/controls/lib/lateral_planner.py:176]: Yeah will do performance at the end. Still quite a few things that are not done. I will write to do list

File: selfdrive/controls/lib/longitudinal_planner.py (hunk 7 above)
Comment (HaraldSchafer) [reply selfdrive/controls/lib/longitudinal_planner.py:20]: Tune isn't final yet.

File: selfdrive/controls/lib/longitudinal_planner.py (hunk 8 above)
Comment (HaraldSchafer) [reply selfdrive/controls/lib/longitudinal_planner.py:93]: Yeah this is needed. If you go out of the acceleration bounds for example and then engage, there will be some jumps. Yeah might want it for lateral too.

**Comment by HaraldSchafer** [comment]
> Will we add a configurable longitudinal actuators delay per car?

Yes, but not needed for this PR

**Comment by HaraldSchafer** [comment]
added todo

File: selfdrive/controls/lib/lead_mpc_lib/generator.cpp [hunk 9]
```diff
@@ -22,8 +18,13 @@ int main( )
 
   Control j_ego;
 
-  auto desired = 4.0 + RW(v_ego, v_l);
+  auto ego_stop_time = v_ego/MAX_BRAKE;
+  auto ego_stop_dist = v_ego*ego_stop_time - (MAX_BRAKE * ego_stop_time*ego_stop_time)/2;
```
Comment (grekiki) [inline selfdrive/controls/lib/lead_mpc_lib/generator.cpp:22]: v_ego*v_ego/(2 * MAX_BRAKE) could be simpler, also gets rid of stop_time

File: selfdrive/controls/lib/lateral_planner.py (hunk 6 above)
Comment (jmayes41) [reply selfdrive/controls/lib/lateral_planner.py:173]: Sounds good to e

**Comment by pd0wm** [comment]
Replaced by https://github.com/commaai/openpilot/pull/22380
